### PR TITLE
Signature handler component for consensus

### PR DESCRIPTION
**Pull request by ssd04**
Added a separate signature handler component for consensus, it is basically a wrapper over the new stateless multisigner from crypto. It keeps the associated public keys and signature shares used for consensus.
- unit tests
- temp elrond-go-crypto version

### CODE REVIEW

File: consensus/mock/consensusDataContainerMock.go [hunk 1]
```diff
@@ -230,6 +231,16 @@ func (ccm *ConsensusCoreMock) SetNodeRedundancyHandler(nodeRedundancyHandler con
 	ccm.nodeRedundancyHandler = nodeRedundancyHandler
 }
 
+// SignatureHandler -
+func (ccm *ConsensusCoreMock) SignatureHandler() consensus.SignatureHandler {
+	return ccm.signatureHandler
+}
+
+// SetSignatureHandler -
```
Comment (AdoAdoAdo) [inline consensus/mock/consensusDataContainerMock.go:239]: this Set method is only available on the mock, not in production right?

File: consensus/interface.go [hunk 2]
```diff
@@ -147,3 +147,17 @@ type ScheduledProcessor interface {
 	IsProcessedOKWithTimeout() bool
 	IsInterfaceNil() bool
 }
+
+// SignatureHandler defines the behaviour of a component that handles signatures in consensus
+type SignatureHandler interface {
+	Reset(pubKeys []string) error
+	SetMultiSignerByEpoch(epoch uint32) error
```
Comment (AdoAdoAdo) [inline consensus/interface.go:154]: this could be removed

File: testscommon/cryptoMocks/multiSignerNewStub.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package cryptoMocks
+
+// MultiSignerNewStub implements crypto multisigner
+// TODO: change name
```
Comment (AdoAdoAdo) [inline testscommon/cryptoMocks/multiSignerNewStub.go:4]: solve TODO?

File: consensus/signing/signing_test.go [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+		require.Equal(t, signing.ErrNoPublicKeySet, err)
+	})
+
+	t.Run("failed to get multi signer for epoch zero", func(t *testing.T) {
+		t.Parallel()
+
+		args := createMockArgsSignatureHolder()
+
+		expectedErr := errors.New("expected error")
+		args.MultiSignerContainer = &cryptoMocks.MultiSignerContainerStub{
```
Comment (AdoAdoAdo) [inline consensus/signing/signing_test.go:88]: fix the issues?

File: consensus/spos/bls/subroundSignature.go [hunk 5]
```diff
@@ -177,13 +186,13 @@ func (sr *subroundSignature) receivedSignature(_ context.Context, cnsDta *consen
...
-	currentMultiSigner, err := sr.MultiSignerContainer().GetMultiSigner(sr.Header.GetEpoch())
+	err = sr.SignatureHandler().SetMultiSignerByEpoch(sr.Header.GetEpoch())
 	if err != nil {
-		log.Error("receivedSignature.GetMultiSigner", "error", err.Error())
+		log.Debug("receivedSignature.SetMultiSignerByEpoch", "error", err.Error())
 		return false
 	}
 
-	err = currentMultiSigner.VerifySignatureShare([]byte(node), sr.GetData(), cnsDta.SignatureShare)
+	err = sr.SignatureHandler().VerifySignatureShare(uint16(index), cnsDta.SignatureShare, sr.GetData())
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundSignature.go:195]: similar to my comment above, add epoch as parameter to VerifySignatureShare and remove the SetMultiSignerByEpoch above.

File: consensus/spos/bls/subroundEndRound.go [hunk 6]
```diff
@@ -195,17 +195,14 @@ func (sr *subroundEndRound) doEndRoundJobByLeader() bool {
...
 		log.Error("doEndRoundJobByLeader.GetMultiSigner", "error", err.Error())
 		return false
 	}
 
-	// TODO: get the signatures from consensus state
-	pubKeysSigners, sigShares := sr.getSigningData()
-
 	// Aggregate sig and add it to the block
-	sig, err := multiSigner.AggregateSigs(pubKeysSigners, sigShares)
+	sig, err := sr.SignatureHandler().AggregateSigs(bitmap)
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound.go:205]: you can add the epoch as parameter to AggregateSigs and remove the Set from above.

File: consensus/spos/bls/subroundSignature.go [hunk 7]
```diff
@@ -79,6 +79,17 @@ func (sr *subroundSignature) doSignatureJob(_ context.Context) bool {
 		return false
 	}
 
+	if check.IfNil(sr.Header) {
+		log.Error("doSignatureJob", "error", spos.ErrNilHeader)
+		return false
+	}
+
+	err = sr.SignatureHandler().SetMultiSignerByEpoch(sr.Header.GetEpoch())
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/spos/bls/subroundSignature.go:87]: :+1:

File: consensus/signing/signing.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+	privKey   crypto.PrivateKey
+	sigShares [][]byte
+	aggSig    []byte
+}
+
+type signatureHolder struct {
+	data                 *signatureHolderData
+	mutSigningData       sync.RWMutex
+	singleSigner         crypto.SingleSigner
+	multiSignerContainer cryptoCommon.MultiSignerContainer
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:29]: removed stored multisigner and the associated mutex

File: consensus/spos/bls/subroundSignature.go (hunk 5 above)
Comment (ssd04) [reply consensus/spos/bls/subroundSignature.go:195]: added epoch parameter

File: consensus/spos/bls/subroundEndRound.go (hunk 6 above)
Comment (ssd04) [reply consensus/spos/bls/subroundEndRound.go:205]: added epoch parameter

File: consensus/mock/consensusDataContainerMock.go (hunk 1 above)
Comment (ssd04) [reply consensus/mock/consensusDataContainerMock.go:239]: right :+1:

File: consensus/signing/signing.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	sh.mutSigningData.Lock()
+	defer sh.mutSigningData.Unlock()
+
+	privKey := sh.data.privKey
+
+	data := &signatureHolderData{
+		pubKeys:   pk,
+		privKey:   privKey,
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:113]: done

File: consensus/signing/signing.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+// CreateSignatureShare returns a signature over a message
+func (sh *signatureHolder) CreateSignatureShare(message []byte, selfIndex uint16) ([]byte, error) {
+	if message == nil {
+		return nil, ErrNilMessage
+	}
+
+	sh.mutSigningData.Lock()
+	defer sh.mutSigningData.Unlock()
+
+	privKeyBytes, err := sh.data.privKey.ToByteArray()
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:131]: :+1:

File: consensus/signing/signing.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+	sigSharesSize := uint16(len(args.PubKeys))
+	sigShares := make([][]byte, sigSharesSize)
+	pk, err := convertStringsToPubKeysBytes(args.PubKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	data := &signatureHolderData{
+		pubKeys:   pk,
+		privKey:   args.PrivKey,
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:49]: done

File: testscommon/cryptoMocks/multiSignerNewStub.go (hunk 3 above)
Comment (ssd04) [reply testscommon/cryptoMocks/multiSignerNewStub.go:4]: done

File: consensus/signing/signing.go [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+	return sh.multiSigner.VerifySignatureShare(pubKey, message, sig)
+}
+
+// StoreSignatureShare stores the partial signature of the signer with specified position
+func (sh *signatureHolder) StoreSignatureShare(index uint16, sig []byte) error {
+	// TODO: evaluate verifying if sig bytes is a valid BLS signature
+	if sig == nil {
+		return ErrNilSignature
+	}
+
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:181]: at least right now, that method is not exposed yet, will keep the nil check for now

File: consensus/signing/signing.go [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+// Reset resets the data inside the signature holder component
+func (sh *signatureHolder) Reset(pubKeys []string) error {
+	if pubKeys == nil {
+		return ErrNilPublicKeys
+	}
+
+	sigSharesSize := uint16(len(pubKeys))
+	sigShares := make([][]byte, sigSharesSize)
+	pk, err := convertStringsToPubKeysBytes(pubKeys)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:101]: renamed

File: consensus/signing/signing.go [hunk 14]
```diff
@@ -43,40 +39,31 @@ func NewSignatureHolder(args ArgsSignatureHolder) (*signatureHolder, error) {
...
 
 func checkArgs(args ArgsSignatureHolder) error {
-	if check.IfNil(args.SingleSigner) {
-		return ErrNilSingleSigner
-	}
 	if check.IfNil(args.MultiSignerContainer) {
 		return ErrNilMultiSignerContainer
 	}
-	if check.IfNil(args.PrivKey) {
+	if args.PrivKeyBytes == nil {
```
Comment (AdoAdoAdo) [inline consensus/signing/signing.go:66]: can you check the len instead? 
it can be len 0 but not nil

File: consensus/signing/signing.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply consensus/signing/signing.go:176]: Alright.
The core verification from VerifySigBytes is anyway called on signatures aggregation (sigBytesToSig) and the aggregation will fail if any of the signatures is not a valid (format) signature. 

I would at least change the nil check to a len(sig)>0 check

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: consensus/signing/signing.go (hunk 14 above)
Comment (ssd04) [reply consensus/signing/signing.go]: done

File: consensus/signing/signing.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:176]: :+1:

**Review by AdoAdoAdo — APPROVED** [review]

File: consensus/spos/bls/subroundSignature.go [hunk 15]
```diff
@@ -68,20 +68,23 @@ func (sr *subroundSignature) doSignatureJob(_ context.Context) bool {
 	if !sr.CanDoSubroundJob(sr.Current()) {
 		return false
 	}
-	if check.IfNil(sr.Header) {
+	if sr.Header == nil {
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundSignature.go:71]: is this change needed?

File: consensus/spos/bls/subroundSignature.go [hunk 16]
```diff
@@ -68,20 +68,23 @@ func (sr *subroundSignature) doSignatureJob(_ context.Context) bool {
...
 
-	multiSigner, err := sr.MultiSignerContainer().GetMultiSigner(sr.Header.GetEpoch())
+	selfIndex, err := sr.SelfConsensusGroupIndex()
 	if err != nil {
-		log.Error("doSignatureJob.GetMultiSigner", "error", err.Error())
+		log.Debug("doSignatureJob.SelfConsensusGroupIndex: not in consensus group")
+		return false
+	}
+
+	if check.IfNil(sr.Header) {
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundSignature.go:82]: duplicate with `L71-L74`

File: consensus/signing/signing.go [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+	return multiSigner.VerifySignatureShare(pubKey, message, sig)
+}
+
+// StoreSignatureShare stores the partial signature of the signer with specified position
+func (sh *signatureHolder) StoreSignatureShare(index uint16, sig []byte) error {
+	if len(sig) == 0 {
+		return ErrInvalidSignature
+	}
+
+	sh.mutSigningData.Lock()
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:177]: here there is an update, so we need `Lock`

File: consensus/signing/signing.go [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+	}
+	if len(args.PubKeys) == 0 {
+		return ErrNoPublicKeySet
+	}
+
+	return nil
+}
+
+// Create generated a signature holder component and initializes corresponding fields
+func (sh *signatureHolder) Create(pubKeys []string, index uint16) (*signatureHolder, error) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go:80]: right, forgot to remove it after refactoring

File: consensus/signing/signing.go [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+	return sigShareBytes, nil
+}
+
+// VerifySignatureShare will verify the signature share based on the specified index
+func (sh *signatureHolder) VerifySignatureShare(index uint16, sig []byte, message []byte, epoch uint32) error {
+	if len(sig) == 0 {
+		return ErrInvalidSignature
+	}
+
+	sh.mutSigningData.Lock()
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go]: done

File: consensus/signing/signing.go [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+
+	if sh.data.sigShares[index] == nil {
+		return nil, ErrNilElement
+	}
+
+	return sh.data.sigShares[index], nil
+}
+
+// not concurrent safe, should be used under RLock mutex
+func (sh *signatureHolder) isIndexInBitmap(index uint16, bitmap []byte) error {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go]: changed to return a bool

File: consensus/signing/signing.go [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+	defer sh.mutSigningData.Unlock()
+
+	maxFlags := len(bitmap) * 8
+	flagsMismatch := maxFlags < len(sh.data.pubKeys)
+	if flagsMismatch {
+		return ErrBitmapMismatch
+	}
+
+	pubKeys := make([][]byte, 0)
+	for i := range sh.data.pubKeys {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go]: done

File: consensus/signing/signing.go [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+		err := sh.isIndexInBitmap(uint16(i), bitmap)
+		if err != nil {
+			continue
+		}
+
+		pubKeys = append(pubKeys, sh.data.pubKeys[i])
+	}
+
+	multiSigner, err := sh.multiSignerContainer.GetMultiSigner(epoch)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go]: done

File: consensus/signing/signing_test.go [hunk 23]
```diff
@@ -0,0 +1,540 @@
...
+		t.Parallel()
+
+		args := createMockArgsSignatureHolder()
+		signer, err := signing.NewSignatureHolder(args)
+		require.Nil(t, err)
+		require.False(t, check.IfNil(signer))
+	})
+}
+
+func TestCreate(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing_test.go]: done

File: consensus/spos/bls/subroundSignature.go (hunk 15 above)
Comment (ssd04) [reply consensus/spos/bls/subroundSignature.go]: no, reverted

File: consensus/signing/signing.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply consensus/signing/signing.go:177]: correct, my bad

**Review by sstanculeanu — APPROVED** [review]

File: consensus/signing/signing_test.go [hunk 24]
```diff
@@ -0,0 +1,540 @@
...
+
+		args := createMockArgsSignatureHolder()
+
+		signer, _ := signing.NewSignatureHolder(args)
+		err := signer.Reset([]string{"pubKey1", "pubKey2"})
+		require.Nil(t, err)
+	})
+}
+
+func TestCreateSignatureShare(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing_test.go]: added more unit tests

**Review by sstanculeanu — DISMISSED** [review]

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: consensus/signing/signing.go [hunk 25]
```diff
@@ -270,28 +266,27 @@ func (sh *signatureHolder) Verify(message []byte, bitmap []byte, epoch uint32) e
...
 	if flagsMismatch {
 		return ErrBitmapMismatch
 	}
 
+	multiSigner, err := sh.multiSignerContainer.GetMultiSigner(epoch)
+	if err != nil {
+		return err
+	}
+
 	pubKeys := make([][]byte, 0)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/signing/signing.go]: done

**Review by AdoAdoAdo — APPROVED** [review]
